scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py

```diff
@@ -0,0 +1,131 @@
+agreement. This can be used to, for example, let all suppliers know their
+services are live.
+
+Before using this script you should check that it has all the template
+variables you need for your Notify template, and run it using `--dry-run` to
+make sure it emails all the users you expect (and only them).
+
+Uses the Notify API to send the email. This script *should not* resend emails.
+
+Usage:
+    scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py [options]
+         [--supplier-id=<id> ... | --supplier-ids-from=<file>]
+         <stage> <framework> <notify_api_key> <notify_template_id> <content_path>
+
+Example:
+    scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py \
+        preview g-cloud-11 $NOTIFY_API_KEY '98a9754d-9310-436a-9e7a-cc7504ecd92d' ../digitalmarketplace-frameworks
+
+Parameters:
+    <stage>                     Environment to run script against.
+    <framework>                 Slug of framework to run script against.
+    <notify_api_key>            API key for GOV.UK Notify.
+    <notify_template_id>        Template to send.
+    <content_path>              Path to digitalmarketplace-frameworks repository.
+
+Options:
+    --supplier-id=<id>          ID(s) of supplier(s) to email.
+    --supplier-ids-from=<file>  Path to file containing supplier ID(s), one per line.
+
+    -n, --dry-run               Run script without sending emails.
+
+    -h, --help                  Show this screen.
+"""
+import sys
+from typing import List
+from itertools import chain
+from docopt import docopt
+
+sys.path.insert(0, ".")
+
+from dmapiclient import DataAPIClient
+from dmcontent.content_loader import ContentLoader
+from dmscripts.helpers.email_helpers import scripts_notify_client
+from dmscripts.helpers.auth_helpers import get_auth_token
+from dmscripts.helpers import logging_helpers
+from dmscripts.helpers.supplier_data_helpers import get_supplier_ids_from_args
+from dmutils.email.helpers import hash_string
+from dmutils.env_helpers import get_api_endpoint_from_stage
+
+
+def get_supplier_ids_signed(api_client: DataAPIClient, framework_slug: str) -> List[int]:
+    """
+    Get a list of supplier IDs who are on `framework_slug` and have signed the framework agreement
+    """
+    return [supplier["supplierId"] for supplier in
+            api_client.find_framework_suppliers_iter(framework_slug, agreement_returned=True, with_declarations=False)
+            if supplier["onFramework"]]
+
+
+def get_email_addresses_for_supplier(api_client: DataAPIClient, supplier_id: int) -> List[str]:
+    """Get the email addresses for each user belonging to `supplier_id`"""
+    supplier_users = api_client.find_users_iter(supplier_id=supplier_id, personal_data_removed=False)
+    return [user["emailAddress"] for user in supplier_users if user["active"]]
+
+
+def get_previous_framework(framework: dict) -> dict:
+    """Find the framework iteration that comes before `framework`
```

Comment:
Is this true? It looks to me like it's finding the second newest framework iteration in this family. Usually, that will be the same thing, but not necessarily.

Given the short amount of time we've got, I'm not sure it's worth working out how to do this properly, but some documentation of the pitfalls might be handy.

Reply:
+1

Reply:
Added some notes